cmd/mt-whisper-importer-reader/main.go

```diff
@@ -0,0 +1,328 @@
+		"Chunks that have not completed their chunk span should be skipped",
+	)
+	orgId = flag.Int(
+		"orgid",
+		1,
+		"Organization ID the data belongs to ",
+	)
+	whisperDirectory = flag.String(
+		"whisper-directory",
+		"/opt/graphite/storage/whisper",
+		"The directory that contains the whisper file structure",
+	)
+	readArchivesStr = flag.String(
+		"read-archives",
+		"*",
+		"Comma separated list of positive integers or '*' for all archives",
+	)
+	chunkSpans   []uint32
+	readArchives map[int]struct{}
+	printLock    sync.Mutex
+)
+
+func main() {
+	flag.Parse()
+
+	for _, chunkSpanStrSplit := range strings.Split(*chunkSpanStr, ":") {
+		chunkSpans = append(chunkSpans, dur.MustParseUNsec("chunkspan", chunkSpanStrSplit))
+	}
+
+	if *readArchivesStr != "*" {
+		readArchives = make(map[int]struct{})
+		for _, archiveIdStr := range strings.Split(*readArchivesStr, ",") {
+			archiveId, err := strconv.Atoi(archiveIdStr)
+			if err != nil {
+				throwError(fmt.Sprintf("Invalid archive id %s: %s", archiveIdStr, err))
+			}
+			readArchives[archiveId] = struct{}{}
+		}
+	}
+
+	fileChan := make(chan string)
+
+	wg := &sync.WaitGroup{}
+	for i := 0; i < *threads; i++ {
+		go processFromChan(fileChan, wg)
+	}
+
+	getFileListIntoChan(fileChan, wg)
+	wg.Wait()
+}
+
+func throwError(msg string) {
+	msg = fmt.Sprintf("%s\n", msg)
+	if *exitOnError {
+		panic(msg)
+	} else {
+		printLock.Lock()
+		fmt.Fprintln(os.Stderr, msg)
+		printLock.Unlock()
+	}
+}
+
+func log(msg string) {
+	if *verbose {
+		printLock.Lock()
+		fmt.Println(msg)
+		printLock.Unlock()
+	}
+}
+
+func processFromChan(files chan string, wg *sync.WaitGroup) {
+	client := &http.Client{}
+
+	for file := range files {
+		fd, err := os.Open(file)
+		if err != nil {
+			throwError(fmt.Sprintf("ERROR: Failed to open whisper file '%s': %s\n", file, err))
+			continue
+		}
+		w, err := whisper.OpenWhisper(fd)
+		if err != nil {
+			throwError(fmt.Sprintf("ERROR: Failed to open whisper file '%s': %s\n", file, err))
+			continue
+		}
+
+		log(fmt.Sprintf("Processing file %s", file))
+		b, err := getMetric(w, file).Encode()
+		if err != nil {
+			throwError(fmt.Sprintf("%s", err))
+		}
+
+		req, err := http.NewRequest("POST", *httpEndpoint, b)
+		if err != nil {
+			throwError(fmt.Sprintf("Cannot send request to http endpoint %s: %s", *httpEndpoint, err))
+		}
+		req.Header.Set("Content-Type", "application/json")
+		req.Header.Set("Content-Encoding", "gzip")
+
+		_, err = client.Do(req)
+		if err != nil {
+			throwError(fmt.Sprintf("Error sending request to http endpoint %s: %s", *httpEndpoint, err))
+		}
+
+		wg.Done()
+	}
+}
+
+// generate the metric name based on the file name and given prefix
+func getMetricName(file string) string {
+	for {
+		if file[:1] == "/" {
+			file = file[1:]
+		} else {
+			break
+		}
+	}
+	splits := strings.Split(file, "/")
+	leafNode := strings.Split(splits[len(splits)-1], ".")
+	splits[len(splits)-1] = leafNode[0]
+	return *namePrefix + strings.Join(splits, ".")
```

Comment:
this function needs some explaining. when i try it with some paths i end up with double dots and a missing prefix.
(https://play.golang.org/p/bZy9IR2pjY) probably because i'm not aware of certain conditions that should always be true. should the filename always start with `/` for example? this function should document what kind of formats it expects, and what it emits. unit tests would be good too.